src/ferc_xbrl_extractor/datapackage.py

```diff
@@ -0,0 +1,409 @@
+                if resource:
+                    resources.append(resource)
+
+        return cls(resources=resources)
+
+    def get_fact_tables(self):
+        """Get FactTables from datapackage."""
+        return {
+            resource.name: FactTable(resource.schema_, resource.get_period_type())
+            for resource in self.resources
+        }
+
+
+class FactTable(object):
+    """
+    Class to handle constructing a dataframe from an XBRL fact table.
+
+    Structure of the dataframe is defined by the XBRL taxonomy. Facts and contexts
+    parsed from an individual XBRL filing are then used to populate the dataframe
+    with relevant data.
+    """
+
+    def __init__(self, schema: Schema, period_type: str):
+        """Create FactTable and prepare for constructing dataframe."""
+        self.schema = schema
+        # Map column names to function to convert parsed values
+        self.columns = {
+            field.name: CONVERT_DTYPES[field.type_] for field in schema.fields
+        }
+        self.axes = [name for name in schema.primary_key if name.endswith("Axis")]
+        self.instant = period_type == "instant"
+        self.logger = get_logger(__name__)
+
+    def construct_dataframe(
+        self, facts: Dict[str, Fact], contexts: Dict[str, Context], filing_name: str
+    ) -> pd.DataFrame:
+        """
+        Construct dataframe from facts and contexts extracted from filing.
+
+        Args:
+            facts: Dictionary mapping context id's to facts.
+            contexts: Dictionary mapping context id's to contexts.
+            filing_name: Name of current filing.
+        """
+        # Filter contexts to only those relevant to table
+        contexts = {
+            c_id: context
+            for c_id, context in contexts.items()
+            if context.check_dimensions(self.axes)
+        }
+
+        # Get the maximum number of rows that could be in table and allocate space
+        max_len = len(contexts)
+
+        # Allocate space for columns
+        df = {
+            field.name: pd.Series([pd.NA] * max_len, dtype=FIELD_TO_PANDAS[field.type_])
+            for field in self.schema.fields
+        }
+
+        # Loop through contexts and get facts in each context
+        for i, (c_id, context) in enumerate(contexts.items()):
+            if self.instant != context.period.instant:
+                continue
+
+            row = {fact.name: fact.value for fact in facts[c_id] if fact.name in df}
+
+            if row:
+                row.update(contexts[c_id].get_context_ids(filing_name))
+
+                for key, val in row.items():
+                    # Try to parse value from XBRL if not possible it will remain NA
+                    try:
+                        df[key][i] = self.columns[key](val)
+                    except ValueError:
+                        self.logger.warning(
+                            f"Could not parse {val} as {self.columns[key]} in column {key}"
+                        )
```

Comment:
Some comments explaining what each step does here would be helpful.

Reply:
Added some comments, let me know if they help!

Reply:
Looks good thanks!

Reply:
Looks good thanks!